### PR TITLE
👑 Private keys import and accounts backup 

### LINKED ISSUE #3070: Update keyring service's `importKeyring` method to accept private keys.
We want this method to be able to support _both_ importing a keyring via seed phrase and importing a keyring via private key.

Accomplishing this will proably involve making some changed changes to the constructor of HDKeyring in the [hd-keyring](https://github.com/tallyhowallet/hd-keyring) repository and digging into ethers to figure out exactly which methods we need to use in order to create an HD Wallet via a private key.

For this ticket to be closed - we should be able to:
- [ ] Import a keyring via a private key and then successfully sign transactions and/or messages using that keyring
- [ ] Be able to derive child addresses from the private key via the default Ethereum derivation path.

### PR DESCRIPTION
**Pull request by jagodarybacka**
### About this PR
This PR will contain all changes for features:
- [x] allow importing private keys in the keyring service (initial scope of this PR)
- [x] import private keys with plain text (https://github.com/tahowallet/extension/pull/3119)
- [x] import private keys with JSON (https://github.com/tahowallet/extension/pull/3177)
- [x] backup account with mnemonic and private key export (https://github.com/tahowallet/extension/pull/3252)
---

## Initial scope 
Resolves #3070

### What

Handle importing wallets with private keys in the keyring service:
* save pk imported accounts in the encrypted vault
* handle serialization and deserialization of pk imported wallets
* handle signing messages with both keyrings and pk wallets
* handle communication between service and redux store about pk wallets
* fix types and field names in the redux store

Cleanup and add more tests:
- move existing keyring integration tests to the right folder
- write more unit tests to cover private key imported wallets
- cleanup some duplicated tests
- add util functions to the test's `factories`
- add function to mock local storage

### Testing
- make sure unit tests and E2E are passing
- test manually adding accounts and signing transactions

Latest build: [extension-builds-3089](https://github.com/tahowallet/extension/suites/14105826655/artifacts/789186059) (as of Thu, 06 Jul 2023 12:56:07 GMT).

### CODE REVIEW

**Comment by Shadowfiend** [comment]
> ⚠️ Please do not merge without code owner's approval! ⚠️

Btw this will be enforced by GitHub, as long as the target branch is `main`.

File: background/services/keyring/index.ts [hunk 1]
```diff
@@ -25,6 +26,19 @@ import logger from "../../lib/logger"
...
 export const MAX_OUTSIDE_IDLE_TIME = 60 * MINUTE
 
+export type WalletData = {
+  type: KeyringTypes.singleSECP
+  id: string | null
+  path: null
+  addresses: [string]
+}
+
+export type SerializedWallet = {
```
Comment (Shadowfiend) [inline background/services/keyring/index.ts:36]: Maybe let's put this in `persistence`? The de facto interaction style of the repo is that `index` is considered to be exporting types and data for consumption outside of the directory, so having this exported here means we expect to export access to the serialized wallet including private key. In practice, we don't currently intend to do that (unless we add private key exporting, but at that point the thing we expose here will likely already be encrypted, while this data type will carry unencrypted private keys).

Another option is just not to export the type, maybe—unclear if it's referenced elsewhere right now.

**Review by Shadowfiend — CHANGES_REQUESTED** [review]
Left some higher-level architectural thoughts on how to reconsider the addition of new methods and redux actions by instead adding new variants to the existing data structures, and reconsidering private keys as “simple” keyrings with only one key and no derivation capabilities.

File: background/services/keyring/encryption.ts [hunk 2]
```diff
@@ -140,7 +140,7 @@ export async function encryptVault<V>(
   return {
     salt,
     initializationVector: bufferToBase64(initializationVector),
-    cipherText: bufferToBase64(cipherText),
+    cipherText: bufferToBase64(cipherText as Uint8Array),
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/encryption.ts:143]: my VSC typescript plugin was highlighting that, but no error in the console 🤔 may be a bug in VSC, will check

**Comment by jagodarybacka** [comment]
- did some renames and other minor improvements (naming ideas still appreciated)
- created a single point of import both HD wallets and private keys
- right now I cannot see how merging HD keyrings and pk wallets into single array will be beneficial, both in the service and in the redux as they have different methods and properties. I need to merge it to the other PR to see how it will behave with selectors etc.

Moving this to draft for now, will notify when it is ready.

**Review by Shadowfiend — COMMENTED** [review]
Ok, so first as I've been digesting this, a categorical issue here on my end: we have mixed the concept of an HDKeyring, which is a keyring that has a mnemonic and can derive many addresses, with the general concept of a Taho keyring, which is here treated as where all internal accounts are managed. Conceptualized in the physical world, the Taho keyring service is a sort of keyring of keys AND keyrings once we finish this work.

My suggestion here is that we rename `KeyringService`, as well as the associated slice and the associated signer subcomponents, to be `InternalSignerService`, `internal-signer` slice, `SignerInternal` for the components, etc. The idea here is that the commonality among all of these isn't that they're keyrings anymore, it's that they are signers whose material we manage internally.

This would also entail renaming the types in `keyring/index.ts` that here are `Signer*` to `InternalSigner*`, which distinguishes them from the more generic types we're using for the signing slice and signing service.

It also should be fuel for us to distinguish between “this is a signing request for an internal signer”, which is a routing directive to go to the internal signer service, from the metadata used by the internal signer service to figure out if it needs a private key or a keyring to do the work (e.g. https://github.com/tahowallet/extension/blob/7784ccce5e9cd3ae9547c17e1ab3542793abc0ba/background/services/signing/index.ts#L141-L149 indicates the signing service knows something it really doesn't need to—it really just needs to know this is an internal account). This is also a generic top-level decision maker for whether we need to check for the keyring being locked/unlocked, etc.

I realize this is a pretty big suggestion/request, so posting it so we can ponder it a little while I do other parts of the code review.

File: ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    setIsImporting(false)
+    setIsImported(true)
+  }, [dispatch, file, password, setIsImporting])
+
+  useEffect(() => {
+    if (keyringImportStatus === "failed" && isImported) {
+      setHasError(true)
+      setIsImported(false)
+    }
```
Comment (Shadowfiend) [inline ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx:68]: This is feeling like debt piling on some debt decisions we made back in the day in terms of how we handle error flow.

I don't think it's blocking at all, but my suggestion for both keyring and private key (and JSON) import is to await an async thunk that returns success/failure, and handle that directly here, vs the current approach of managing this weird `importing` field that has `false`, `"pending"`, and `"done"` as values, and triggering effects based on them. The state update loop ends up being quite hard to follow and even harder to debug vs saying `await dispatch(import)` followed by `if (success) { ... } else { ... }`.

File: ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+    if (keyringImportStatus === "done" && isImported) {
+      dispatch(sendEvent(OneTimeAnalyticsEvent.ONBOARDING_FINISHED))
```
Comment (Shadowfiend) [inline ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx:71]: Not blocking: is there a centralized place we could do this? e.g. as an effect on https://github.com/tahowallet/extension/blob/fa1552e74351ebb865ff70568e3832e1a27f8e3a/ui/pages/Onboarding/Tabbed/Done.tsx#L74 or similar?

**Review by Shadowfiend — COMMENTED** [review]
Couple more questions, still loading heh ⌛

File: background/services/keyring/index.ts [hunk 5]
```diff
@@ -31,28 +32,73 @@ export type Keyring = {
...
+  password: string
+}
+export type SignerRawWithType =
+  | SignerRawPrivateKey
+  | SignerRawHDKeyring
+  | SignerRawJSONPrivateKey
+
+type SignerHDKeyring = { type: SignerTypes.keyring; signer: HDKeyring }
+type SignerPrivateKey = { type: SignerTypes.privateKey; signer: Wallet }
+type SignerWithType = SignerPrivateKey | SignerHDKeyring
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:86]: The `SigerRaw...` types are needed before we create the actual signer entity that is either `HDKeyring` or ether's `Wallet` types. So `SigerRaw...`s are holding metadata needed to create signers objects.

If you prefer other names we can go with something else instead of using `Raw`. Maybe `Metadata`?

File: ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx (hunk 4 above)
Comment (jagodarybacka) [reply ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx:71]: I believe we want to exclude onboarding with read only addresses. Other than that yeah we can refactor it later.

File: ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx (hunk 3 above)
Comment (jagodarybacka) [reply ui/pages/Onboarding/Tabbed/ImportPrivateKeyJSON.tsx:68]: Done in https://github.com/tahowallet/extension/pull/3309

File: background/services/keyring/index.ts [hunk 6]
```diff
@@ -31,28 +32,73 @@ export type Keyring = {
...
+type SignerRawPrivateKey = { type: SignerTypes.privateKey; privateKey: string }
+type SignerRawJSONPrivateKey = {
+  type: SignerTypes.jsonFile
+  jsonFile: string
+  password: string
+}
+export type SignerRawWithType =
+  | SignerRawPrivateKey
+  | SignerRawHDKeyring
+  | SignerRawJSONPrivateKey
```
Comment (hyphenized) [inline background/services/keyring/index.ts:82]: I think we've abstracted this too early, we don't need the type guards yet either as we're only using them in a single place, and can already discriminate by the `type` property before passing each `import...` function it's required data rather than passing again the same object.

What do you think about naming this `RawSignerWithType`? If we're keeping the type aliases I'd suggest following the same naming pattern as well i.e. `RawSignerPrivateKey`, `RawSignerHDKeyring`, and so on.

**Review by hyphenized — CHANGES_REQUESTED** [review]
Implementation looks good so far! Gave a first pass and left a few comments, will follow up shortly.

File: background/redux-slices/keyrings.ts [hunk 7]
```diff
@@ -12,17 +18,19 @@ type KeyringToVerify = {
 
 export type KeyringsState = {
   keyrings: Keyring[]
-  keyringMetadata: {
-    [keyringId: string]: KeyringMetadata
+  privateKeys: PrivateKey[]
```
Comment (0xDaedalus) [inline background/redux-slices/keyrings.ts:21]: Small nit here that the type name makes it seem like we're storing Private Keys in redux state

File: background/redux-slices/keyrings.ts [hunk 8]
```diff
@@ -76,26 +79,28 @@ const keyringsSlice = createSlice({
...
           keyrings: Keyring[]
-          keyringMetadata: { [keyringId: string]: KeyringMetadata }
+          metadata: { [keyringId: string]: SignerMetadata }
         }
       }
     ) => {
       // When the keyrings are locked, we receive updateKeyrings with an empty
       // list as the keyring service clears the in-memory keyrings. For UI
       // purposes, however, we want to continue tracking the keyring metadata,
       // so we ignore an empty list if the keyrings are locked.
```
Comment (0xDaedalus) [inline background/redux-slices/keyrings.ts:94]: We should update the comment here

File: background/services/keyring/index.ts [hunk 9]
```diff
@@ -430,22 +554,52 @@ export default class KeyringService extends BaseService<Events> {
...
     this.emitKeyrings()
   }
 
+  async exportPrivateKey(address: HexString): Promise<string | null> {
+    this.requireUnlocked()
+
+    try {
+      const privateKeyWallet = await this.#findPrivateKey(address)
+      return privateKeyWallet.privateKey
+    } catch (e) {
```
Comment (0xDaedalus) [inline background/services/keyring/index.ts:586]: IS there any value in logging the error here?

File: background/services/keyring/index.ts [hunk 10]
```diff
@@ -430,22 +554,52 @@ export default class KeyringService extends BaseService<Events> {
...
+  }
+
+  async exportMnemonic(address: HexString): Promise<string | null> {
+    this.requireUnlocked()
+
+    try {
+      const keyring = await this.#findKeyring(address)
+      const { mnemonic } = await keyring.serialize()
+      return mnemonic
+    } catch (e) {
```
Comment (0xDaedalus) [inline background/services/keyring/index.ts:598]: Same as above

File: background/services/keyring/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:86]: Changed in https://github.com/tahowallet/extension/pull/3331

File: background/redux-slices/selectors/accountsSelectors.ts [hunk 11]
```diff
@@ -317,6 +318,9 @@ const getAccountType = (
   if (signerTypeToAccountType[signer.type] === "ledger") {
     return AccountType.Ledger
   }
+  if (signerTypeToAccountType[signer.type] === "privateKey") {
+    return AccountType.PrivateKey
+  }
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/redux-slices/selectors/accountsSelectors.ts:323]: https://github.com/tahowallet/extension/pull/3331/commits/d82d83aa895e8efb68e86512178273cb7a50b8f5

File: background/redux-slices/accounts.ts [hunk 12]
```diff
@@ -26,6 +26,7 @@ import { convertFixedPoint } from "../lib/fixed-point"
  */
 export const enum AccountType {
   ReadOnly = "read-only",
+  PrivateKey = "privateKey",
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/redux-slices/accounts.ts:29]: https://github.com/tahowallet/extension/pull/3331/files#diff-41888dfa5421a9149c733a79b8c8b7f83ca5c79b12b4346025e14f19ab31cb0dR28

File: ui/public/images/add_wallet/import_priv_key.svg
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply ui/public/images/add_wallet/import_priv_key.svg:1]: https://github.com/tahowallet/extension/pull/3331/commits/b67a8a91138b0bb3d61872dc6d3269c914f70d9b

File: ui/pages/Onboarding/Tabbed/Routes.ts [hunk 13]
```diff
@@ -4,6 +4,7 @@ const OnboardingRoutes = {
   LEDGER: "/onboarding/ledger",
   SET_PASSWORD: "/onboarding/set-password",
   IMPORT_SEED: "/onboarding/import-seed",
+  IMPORT_PRIVATE_KEY: "/onboarding/import-priv-key",
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply ui/pages/Onboarding/Tabbed/Routes.ts:7]: https://github.com/tahowallet/extension/pull/3331/commits/30e5fc869ad51b561e1ea8d63778d7140d5bc2b2

File: background/services/keyring/index.ts [hunk 14]
```diff
@@ -326,46 +394,91 @@ export default class KeyringService extends BaseService<Events> {
...
    * @returns The string ID of the new keyring.
    */
-  async importKeyring(
-    mnemonic: string,
-    source: "import" | "internal",
-    path?: string
-  ): Promise<string> {
-    this.requireUnlocked()
+
+  async #importKeyring(signerRaw: SignerRawHDKeyring): Promise<string | null> {
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:427]: [Comments updated](https://github.com/tahowallet/extension/pull/3331/files#diff-a8aee172609c7e4b51c07f3ba53e2e0c24786f91a74f1b03d2cc8cbbe3881e09R418)

File: background/services/chain/index.ts [hunk 15]
```diff
@@ -916,7 +916,7 @@ export default class ChainService extends BaseService<Events> {
   }
 
   async addAccountToTrack(addressNetwork: AddressOnNetwork): Promise<void> {
-    const source = await this.keyringService.getKeyringSourceForAddress(
+    const source = await this.keyringService.getSignerSourceForAddress(
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/chain/index.ts:919]: https://github.com/tahowallet/extension/pull/3331/commits/c33ce42f86c1feda2fb0e40aca0627a9718e5eef

File: background/services/keyring/index.ts [hunk 16]
```diff
@@ -490,13 +687,22 @@ export default class KeyringService extends BaseService<Events> {
     const { address: account, network } = addressOnNetwork
 
     // find the keyring using a linear search
-    const keyring = await this.#findKeyring(account)
+    const signerWithType = await this.#findSigner(account)
 
     // ethers has a looser / slightly different request type
     const ethersTxRequest = ethersTransactionFromTransactionRequest(txRequest)
 
+    let signed: string
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:703]: `signedTx` is declared below but https://github.com/tahowallet/extension/pull/3331/commits/3e436e73a930e57a8954735b6832bd9842f18708

File: background/redux-slices/keyrings.ts (hunk 7 above)
Comment (jagodarybacka) [reply background/redux-slices/keyrings.ts:22]: Yeah but idk what is the better name as the feature that distinguishes these wallets is that they are imported with private keys 🥲

File: background/redux-slices/keyrings.ts (hunk 8 above)
Comment (jagodarybacka) [reply background/redux-slices/keyrings.ts:110]: https://github.com/tahowallet/extension/pull/3331/commits/caa734e184df9aadbe0534787ad10ae1b80acfa6

File: background/services/keyring/index.ts (hunk 9 above)
Comment (jagodarybacka) [reply background/services/keyring/index.ts:594]: Yeah logging them won't hurt https://github.com/tahowallet/extension/pull/3331/commits/72eb623fcc743e181f1ea1938398bef69387d771

File: background/services/keyring/index.ts [hunk 17]
```diff
@@ -470,11 +639,39 @@ export default class KeyringService extends BaseService<Events> {
       kr.getAddressesSync().includes(normalizeEVMAddress(account))
     )
     if (!keyring) {
-      throw new Error("Address keyring not found.")
+      throw new Error(`Keyring not found for address ${account}.`)
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:650]: https://github.com/tahowallet/extension/pull/3331/commits/d9f750ee63ab76e3dc6e58934e96970a3160ec6b there is still some cleanup to do about what is being logged and what it just thrown as error, will do it later

File: background/services/keyring/index.ts [hunk 18]
```diff
@@ -326,46 +394,91 @@ export default class KeyringService extends BaseService<Events> {
...
+    return address
+  }
+
+  async #importPrivateKey(
+    signerRaw: SignerRawPrivateKey
+  ): Promise<string | null> {
+    const { privateKey } = signerRaw
+    const newWallet = new Wallet(privateKey)
+    const normalizedAddress = normalizeEVMAddress(newWallet.address)
+    // TODO: check if this wallet already exists
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:450]: https://github.com/tahowallet/extension/pull/3331/commits/6357ca1757b98bc8c89b680b7c9a75d84df12203

**Review by hyphenized — COMMENTED** [review]
This is looking great! Gave another pass through the rest of the files, left mostly minor comments.

File: background/services/keyring/index.ts [hunk 19]
```diff
@@ -326,46 +394,91 @@ export default class KeyringService extends BaseService<Events> {
     return { id: newKeyring.id, mnemonic: mnemonic.split(" ") }
   }
 
+  async importSigner(signerRaw: SignerRawWithType): Promise<HexString | null> {
+    this.requireUnlocked()
+    let address: HexString | null
+
+    if (isRawPrivateKey(signerRaw)) {
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:401]: https://github.com/tahowallet/extension/pull/3331/commits/c3778039808abc092a92c1138e75feeec9a5f0ea

File: background/services/keyring/index.ts (hunk 6 above)
Comment (jagodarybacka) [reply background/services/keyring/index.ts:82]: > discriminate by the type property before passing each import... function it's required data rather than passing again the same object

If I understand correctly what you would like to do - it would require to use `as` or other type casting solutions to ensure each import method is getting correct metadata. So idk if this would be cleaner solution than just having separate types here

I renamed them in the other PR already, lmk later if that's better.

File: background/services/keyring/index.ts [hunk 20]
```diff
@@ -31,28 +32,73 @@ export type Keyring = {
...
   keyringID: string
 }
 
-export interface KeyringMetadata {
+export type PrivateKeyAccountSigner = {
+  type: "privateKey"
+  walletID: string
+}
+
+export type SignerMetadata = {
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:51]: https://github.com/tahowallet/extension/pull/3331/commits/2519ed33634a8b76855de2d963ac29c5757af477

File: background/services/keyring/index.ts [hunk 21]
```diff
@@ -646,12 +865,18 @@ export default class KeyringService extends BaseService<Events> {
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:873]: https://github.com/tahowallet/extension/pull/3331/commits/d366873f3339a4d9c6c00b843262195b1b8bc054

File: background/redux-slices/migrations/to-28.ts [hunk 22]
```diff
@@ -0,0 +1,40 @@
+import { PrivateKey } from "../../services/keyring"
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/redux-slices/migrations/to-28.ts:1]: https://github.com/tahowallet/extension/pull/3331/commits/3e1a3fefc5d8c3de133c392d40a27b6a7d6eeff8

File: ui/components/AccountsNotificationPanel/AccountsNotificationPanelAccounts.tsx [hunk 23]
```diff
@@ -51,6 +51,14 @@ type WalletTypeInfo = {
   category: string
 }
 
+const accountTypes = [
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply ui/components/AccountsNotificationPanel/AccountsNotificationPanelAccounts.tsx:60]: https://github.com/tahowallet/extension/pull/3331/commits/a983330a8d07511e0f89af61dd040626ca5b7d11

File: background/services/keyring/index.ts [hunk 24]
```diff
@@ -31,28 +32,73 @@ export type Keyring = {
   path: string | null
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply background/services/keyring/index.ts:32]: https://github.com/tahowallet/extension/pull/3331/commits/bb8f58ecdf77ed3bfe3092c8d6e4e0e74f83695f

File: background/services/keyring/index.ts (hunk 6 above)
Comment (hyphenized) [reply background/services/keyring/index.ts:82]: Sorry, by discriminating by the `type` prop I meant that we can do the following, and remove the type guards:

```ts
	 if (signerMetadata.type === SignerSourceTypes.privateKey) {
      address = this.#importPrivateKey(signerMetadata)
    } else if (signerMetadata.type === SignerSourceTypes.jsonFile) {
      address = await this.#importJSON(signerMetadata)
    } else {
      address = this.#importKeyring(signerMetadata)
    }
```

But it looks that it has already been done in c3778039808abc092a92c1138e75feeec9a5f0ea.

Moving on to the import methods, there doesn't seem to be a need for each of these to require this same `type` property. For instance, `importPrivateKey` only needs to be aware of the private key as a parameter.

File: ui/pages/Onboarding/Tabbed/ImportPrivateKey.tsx [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+import { sendEvent } from "@tallyho/tally-background/redux-slices/ui"
+import { SignerTypes } from "@tallyho/tally-background/services/keyring"
+import { isHexString } from "ethers/lib/utils"
+import React, { ReactElement, useCallback, useState } from "react"
+import { useTranslation } from "react-i18next"
+import SharedButton from "../../../components/Shared/SharedButton"
+import SharedSeedInput from "../../../components/Shared/SharedSeedInput"
+import { useBackgroundDispatch } from "../../../hooks"
+
+type Props = {
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply ui/pages/Onboarding/Tabbed/ImportPrivateKey.tsx:13]: Yeah, I think it is very unlikely that we will ever need to reuse component props types, let's discuss it later.

File: ui/components/AccountsNotificationPanel/AccountsNotificationPanelAccounts.tsx [hunk 26]
```diff
@@ -67,13 +75,39 @@ export const walletTypeDetails: { [key in AccountType]: WalletTypeInfo } = {
...
+      return !existingAccountTypes.includes(AccountType.Internal)
+    case AccountType.PrivateKey:
+      return !(
+        existingAccountTypes.includes(AccountType.Internal) ||
+        existingAccountTypes.includes(AccountType.Imported)
+      )
+    default:
+      throw Error("Unknown account type")
+  }
+}
```
[the remark this answers is not in the document]
Comment (jagodarybacka) [reply ui/components/AccountsNotificationPanel/AccountsNotificationPanelAccounts.tsx:115]: https://github.com/tahowallet/extension/pull/3331/commits/825d6cea68fc2a179831fa35365e80982af47895

**Review by hyphenized — DISMISSED** [review]

**Review by hyphenized — APPROVED** [review]

**Comment by Shadowfiend** [comment]
Let's switch the copy “Private Key Import” on the onboarding screen to “Import private key” to match “Import recovery phrase”

**Review by Shadowfiend — DISMISSED** [review]
All righty, we are good to go here. Waiting on one more round of thoughts from the auditors and then we're going to land this puppy with great fanfare.

**Review by Shadowfiend — APPROVED** [review]
It is time 🏆
